### PR TITLE
[RLlib] Redo: Make TFModelV2 fully modular like TorchModelV2 (soft-deprecate register_variables, unify var names wrt torch).

### PR DESCRIPTION
**Pull request by sven1977**
Redo: Make TFModelV2 fully modular like TorchModelV2 (soft-deprecate register_variables, unify var names).
- Same as this PR: https://github.com/ray-project/ray/pull/13339
- But: Fixed the BatchNormModel example.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/examples/models/batch_norm_model.py [hunk 1]
```diff
@@ -13,13 +13,73 @@
 torch, nn = try_import_torch()
 
 
+class KerasBatchNormModel(TFModelV2):
```
Comment (michaelzhiluo) [inline rllib/examples/models/batch_norm_model.py:16]: Confirmed working now right?

File: rllib/models/tests/test_distributions.py [hunk 2]
```diff
@@ -94,8 +94,7 @@ def test_categorical(self):
 
         inputs = inputs_space.sample()
 
-        for fw, sess in framework_iterator(
-                session=True, frameworks=("tf", "tf2", "torch")):
+        for fw, sess in framework_iterator(session=True):
```
Comment (michaelzhiluo) [inline rllib/models/tests/test_distributions.py:97]: Assuming that by default it iterates through all frameworks, I think there are a lot more tests that need to have the same deletion to reflect this update.

File: rllib/models/tests/test_models.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        my_tf_model = TestTFModel(obs_space, action_space, 5, {},
+                                  "my_tf_model")
+        # Call the model.
+        out, states = my_tf_model({"obs": np.array([obs_space.sample()])})
+        self.assertTrue(out.shape == (1, 5))
+        self.assertTrue(out.dtype == tf.float32)
+        self.assertTrue(states == [])
+        vars = my_tf_model.variables(as_dict=True)
+        self.assertTrue(len(vars) == 6)
+        self.assertTrue("keras_model.dense.kernel:0" in vars)
```
Comment (michaelzhiluo) [inline rllib/models/tests/test_models.py:48]: This test will probably break as soon as keras updates. Idk if there is a better way...

File: rllib/examples/models/batch_norm_model.py (hunk 1 above)
Comment (sven1977) [reply rllib/examples/models/batch_norm_model.py:16]: Just moved it up in the file. It was always there. I rather have users use this model as it's more in line with our ModelV2 coding standards. Just for the PPO case, it doesn't work. However, our batch_norm example choses the correct model class based on the algo used (for PPO, it choses the old-style model).

File: rllib/models/tests/test_distributions.py (hunk 2 above)
Comment (sven1977) [reply rllib/models/tests/test_distributions.py:97]: Probably. But this is a start :) Ideally, we should test with all frameworks everywhere.

File: rllib/models/tests/test_models.py (hunk 3 above)
Comment (sven1977) [reply rllib/models/tests/test_models.py:48]: Yes, one of the points of this PR is actually to get to a state where we have uniform model weights names (across frameworks), at least for our default models. So this is a good start, I think. The only thing I expect to change with keras in the below var-names would be the ":0", maybe. Everything else is controlled by us (TFModelV2 auto-detection and naming mechanism).

File: rllib/models/tests/test_distributions.py (hunk 2 above)
Comment (michaelzhiluo) [reply rllib/models/tests/test_distributions.py:97]: I mean the currently existing tests that use all frameworks, they should be replaced with this new line of code ^^

**Review by michaelzhiluo — APPROVED** [review]

File: rllib/models/tests/test_distributions.py (hunk 2 above)
Comment (sven1977) [reply rllib/models/tests/test_distributions.py:97]: Yes, you are right. Actually, I'm not even sure why I made these fixes above. Not really related to this PR. We should do a separate PR, where we comb through all occurrences of this and fix those.
